skimage/transform/_warps.py

```diff
@@ -717,19 +717,34 @@ def _clip_warp_output(input_image, output_image, mode, cval, clip):
 
     """
     if clip:
-        min_val = input_image.min()
-        max_val = input_image.max()
-
-        preserve_cval = (mode == 'constant' and not
-                         (min_val <= cval <= max_val))
-
-        if preserve_cval:
-            cval_mask = output_image == cval
-
-        np.clip(output_image, min_val, max_val, out=output_image)
-
-        if preserve_cval:
-            output_image[cval_mask] = cval
+        min_val = np.min(input_image)
+        if np.isnan(min_val):
+            # NaNs detected, use NaN-safe min/max
+            min_func = np.nanmin
+            max_func = np.nanmax
+            min_val = min_func(input_image)
+        else:
+            min_func = np.min
+            max_func = np.max
+        max_val = max_func(input_image)
+
+        # Check if cval has been used such that it expands the effective input
+        # range
+        preserve_cval = (mode == 'constant'
+                         and not np.isnan(cval)
+                         and not min_val <= cval <= max_val
+                         and min_func(output_image) <= cval <= max_func(output_image))
```

Comment:
I'm a little bit confused by the (new) logic in this function. So I tested it locally and noticed that commenting these 3 lines has no effect on testing `test_warps.py` at all! So either the test are not covering this function fully or the logic does not work the way you expect it to.

Why is it relevant to preserve `cval` if it's within `min_func(output_image) <= cval <= max_func(output_image)`? 

If `not min_val <= cval <= max_val`, then the calls to `min()` / `max()` within `np.clip(...)` ensure that it is preserved. So I think that is why this line has no effect.

For `not np.isnan(cval)` to matter, the test suite should use `cval=np.nan`.

Reply:
Sorry, yes, there is incomplete test coverage, so I can add more tests.

> Why is it relevant to preserve `cval` if it's within `min_func(output_image) <= cval <= max_func(output_image)`?

My thinking here is that even if `cval` is not in the input range, we don't want to expand the effective input range to include `cval` unless we know that `cval` was actually used in the output image.  That is, if `cval` is not in the output range, `cval` wasn't actually used, so we proceed with clipping to the input range of `min_val` to `max_val`.

Note that it's not possible to test this logic using `rotate()` because that invariably uses `cval` for the output image for non-trivial rotations, while trivial rotations would not result in any pixel values that would need to be clipped.  The test may be able to get away with using `resize(..., mode='constant')`, but may require using `warp()`.

Reply:
Tests added

It turns out that the explicit `not np.isnan(cval)` is redundant because the `min_func(output_image) <= cval <= max_func(output_image)` can only be true if cval is not NaN.

Surprisingly, it turns out that even if you remove both checks, the rest of the new code as written happens to still pass the test for when `cval` is NaN due to a later quirk.  `cval` will be NaN for the `clip()` call, but that doesn't break the clipping because `min(min_val, np.nan)` evaluates to `min_val`.  I say "surprisingly" because not only does `np.min([min_val, np.nan])` evaluate to `np.nan`, `min(np.nan, min_val)` (i.e., swapping the order of the arguments) evaluates to `np.nan` as well.  I wasn't aware of that consequence of comparison operators for `np.nan` always evaluating to `False`: `min(min_val, np.nan) != min(np.nan, min_val)`.